src/main/kotlin/org/rust/ide/annotator/RustHighlightingAnnotator.kt

```diff
@@ -12,7 +12,11 @@ import org.rust.lang.core.psi.util.elementType
 
 // Highlighting logic here should be kept in sync with tags in RustColorSettingsPage
 class RustHighlightingAnnotator : Annotator {
-    override fun annotate(element: PsiElement, holder: AnnotationHolder) = element.accept(object : RustElementVisitor() {
+
+    override fun annotate(element: PsiElement, _holder: AnnotationHolder) {
+        element.accept(object : RustElementVisitor() {
+        var holder = wrap(_holder)
```

Comment:
I don't know what is the best way to handle this, but I see two possible solutions:

* we can add `val target: PsiElement` to `HighlightingInfo`, such that `apply` method would update the `target`'s coloring. 

* we can color the leaf elements directly, using `leaf.parent` to make decisions about what color should be applied. 

I think that these are orthogonal issues:

* using stateless color computation to be able to reuse code for coloring references without "holder context switches"

* deciding, which is best: color leaf elements while visiting them or their parent.

It's interesting that in Go Plugin they seem to simply duplicate some code for coloring references and non references: https://github.com/go-lang-plugin-org/go-lang-idea-plugin/blob/master/src/com/goide/highlighting/GoHighlightingAnnotator.java